vault/terraform/main.tf

```diff
@@ -0,0 +1,206 @@
+      "Principal": {
+        "Service": "ec2.amazonaws.com"
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_role_policy_attachment" "vault" {
+  role       = aws_iam_role.vault.name
+  policy_arn = aws_iam_policy.vault.arn
+}
+
+resource "aws_iam_policy" "vault" {
+  name        = "vault-policy"
+  description = "Policy for Vault backend access"
+
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListBucket",
+        "s3:GetBucketLocation",
+        "s3:ListBucketMultipartUploads",
+        "s3:ListBucketVersions"
+      ],
+      "Resource": "arn:aws:s3:::${aws_s3_bucket.vault_storage.bucket}"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:GetObject",
+        "s3:PutObject",
+        "s3:DeleteObject",
+        "s3:ListMultipartUploadParts"
+      ],
+      "Resource": "arn:aws:s3:::${aws_s3_bucket.vault_storage.bucket}/*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "kms:Decrypt",
+        "kms:Encrypt",
+        "kms:GenerateDataKey*",
+        "kms:DescribeKey"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": "sts:AssumeRole",
+      "Resource": "*"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_user_policy_attachment" "vault" {
+  user       = aws_iam_user.vault.name
+  policy_arn = aws_iam_policy.vault.arn
+  depends_on = [aws_iam_policy.vault]
+}
+
+resource "aws_kms_key" "vault" {
+  description             = "Vault encryption key"
+  enable_key_rotation     = true
+  deletion_window_in_days = 30
+}
+
+resource "aws_kms_alias" "vault" {
+  name          = "alias/vault-key"
+  target_key_id = aws_kms_key.vault.key_id
+}
+
+resource "aws_security_group_rule" "vault_http" {
+  security_group_id = aws_security_group.vault_sg.id
+  type              = "ingress"
+  from_port         = 8200
+  to_port           = 8200
+  protocol          = "tcp"
+  cidr_blocks       = ["0.0.0.0/0"] # Restrict the CIDR block as needed
+}
+
+# Read private key file and assign it to a variable
+data "local_file" "private_key" {
+  filename = var.private_key_path
+}
+
+resource "aws_instance" "vault" {
+  ami                         = data.aws_ami.ubuntu_x86.id # Update with the desired Vault AMI ID
+  instance_type               = var.vault_instance_type    # Update with the desired instance type
+  key_name                    = var.ssh_key_name           # Update with your SSH key pair
+  subnet_id                   = aws_subnet.vault_subnet.id
+  availability_zone           = var.aws_az
+  associate_public_ip_address = true
+
+  # Attach the IAM role to the instance
+  iam_instance_profile = aws_iam_instance_profile.vault.name
+
+  vpc_security_group_ids = [aws_security_group.vault_sg.id]
+
+  tags = {
+    Name = "vault-server"
+  }
+
+  provisioner "remote-exec" {
+  inline = [
+    "export DEBIAN_FRONTEND=noninteractive",
+    "sudo apt update -y",
```

Comment:
Just wondering, do all of these actually need sudo? I'd expect provisioner to run under root already